diff --git a/Cargo.lock b/Cargo.lock
index 5ef6c47..71c8969 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1651,7 +1651,6 @@ dependencies = [
  "serde_json",
  "slab",
  "thiserror 2.0.18",
- "userfaultfd",
  "utils",
  "uuid",
  "vhost",
diff --git a/src/vmm/Cargo.toml b/src/vmm/Cargo.toml
index 9486e00..4e7367b 100644
--- a/src/vmm/Cargo.toml
+++ b/src/vmm/Cargo.toml
@@ -46,7 +46,6 @@ serde = { version = "1.0.228", features = ["derive", "rc"] }
 serde_json = "1.0.149"
 slab = "0.4.12"
 thiserror = "2.0.18"
-userfaultfd = "0.9.0"
 utils = { path = "../utils" }
 uuid = "1.23.1"
 vhost = { version = "0.15.0", features = ["vhost-user-frontend"] }
diff --git a/src/vmm/src/builder.rs b/src/vmm/src/builder.rs
index 45dd386..b147d2e 100644
--- a/src/vmm/src/builder.rs
+++ b/src/vmm/src/builder.rs
@@ -11,7 +11,6 @@ use std::sync::{Arc, Mutex};

 use event_manager::SubscriberOps;
 use linux_loader::cmdline::Cmdline as LoaderKernelCmdline;
-use userfaultfd::Uffd;
 use utils::time::TimestampUs;
 use vm_allocator::AllocPolicy;
 use vm_memory::GuestAddress;
@@ -440,7 +439,6 @@ pub fn build_microvm_from_snapshot(
     event_manager: &mut EventManager,
     microvm_state: MicrovmState,
     guest_memory: Vec<GuestRegionMmap>,
-    uffd: Option<Uffd>,
     seccomp_filters: &BpfThreadMap,
     vm_resources: &mut VmResources,
     clock_realtime: bool,
@@ -500,8 +498,6 @@ pub fn build_microvm_from_snapshot(
     // Restore the boot source config paths.
     vm_resources.boot_source.config = microvm_state.vm_info.boot_source;

-    vm.set_uffd(uffd);
-
     let kvm_vm = Arc::new(vm);
     let vm = Vm::Kvm(kvm_vm.clone());

diff --git a/src/vmm/src/persist.rs b/src/vmm/src/persist.rs
index feb53c0..422cca6 100644
--- a/src/vmm/src/persist.rs
+++ b/src/vmm/src/persist.rs
@@ -6,16 +6,11 @@
 use std::fmt::Debug;
 use std::fs::{File, OpenOptions};
 use std::io::{self, Write};
-use std::mem::forget;
-use std::os::unix::io::AsRawFd;
-use std::os::unix::net::UnixStream;
 use std::path::Path;
 use std::sync::{Arc, Mutex};

 use semver::Version;
 use serde::{Deserialize, Serialize};
-use userfaultfd::{FeatureFlags, Uffd, UffdBuilder};
-use vmm_sys_util::sock_ctrl_msg::ScmSocket;

 #[cfg(target_arch = "aarch64")]
 use crate::arch::aarch64::vcpu::get_manufacturer_id_from_host;
@@ -98,32 +93,6 @@ pub struct MicrovmState {
     pub device_states: DevicesState,
 }

-/// This describes the mapping between Firecracker base virtual address and
-/// offset in the buffer or file backend for a guest memory region. It is used
-/// to tell an external process/thread where to populate the guest memory data
-/// for this range.
-///
-/// E.g. Guest memory contents for a region of `size` bytes can be found in the
-/// backend at `offset` bytes from the beginning, and should be copied/populated
-/// into `base_host_address`.
-#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
-pub struct GuestRegionUffdMapping {
-    /// Base host virtual address where the guest memory contents for this
-    /// region should be copied/populated.
-    pub base_host_virt_addr: u64,
-    /// Region size.
-    pub size: usize,
-    /// Offset in the backend file/buffer where the region contents are.
-    pub offset: u64,
-    /// The configured page size for this memory region.
-    pub page_size: usize,
-    /// The configured page size **in bytes** for this memory region. The name is
-    /// wrong but cannot be changed due to being API, so this field is deprecated,
-    /// to be removed in 2.0.
-    #[deprecated]
-    pub page_size_kib: usize,
-}
-
 /// Errors related to saving and restoring Microvm state.
 #[derive(Debug, thiserror::Error, displaydoc::Display)]
 pub enum MicrovmStateError {
@@ -348,14 +317,13 @@ pub enum RestoreFromSnapshotError {
     /// Failed to build microVM from snapshot: {0}
     Build(#[from] BuildMicrovmFromSnapshotError),
 }
-/// Sub-Error type for [`restore_from_snapshot`] to contain either [`GuestMemoryFromFileError`] or
-/// [`GuestMemoryFromUffdError`] within [`RestoreFromSnapshotError`].
+/// Error type for restoring snapshot guest memory.
 #[derive(Debug, thiserror::Error, displaydoc::Display)]
 pub enum RestoreFromSnapshotGuestMemoryError {
     /// Error creating guest memory from file: {0}
     File(#[from] GuestMemoryFromFileError),
-    /// Error creating guest memory from uffd: {0}
-    Uffd(#[from] GuestMemoryFromUffdError),
+    /// UFFD-backed snapshot restore is not supported.
+    UffdUnsupported,
 }

 /// Loads a Microvm snapshot producing a 'paused' Microvm.
@@ -440,7 +408,7 @@ pub fn restore_from_snapshot(
     let mem_backend_path = &params.mem_backend.backend_path;
     let mem_state = &microvm_state.vm_state.memory;

-    let (guest_memory, uffd) = match params.mem_backend.backend_type {
+    let guest_memory = match params.mem_backend.backend_type {
         MemBackendType::File => {
             if vm_resources.machine_config.huge_pages.is_hugetlbfs() {
                 return Err(RestoreFromSnapshotGuestMemoryError::File(
@@ -448,26 +416,18 @@ pub fn restore_from_snapshot(
                 )
                 .into());
             }
-            (
-                guest_memory_from_file(mem_backend_path, mem_state, track_dirty_pages)
-                    .map_err(RestoreFromSnapshotGuestMemoryError::File)?,
-                None,
-            )
+            guest_memory_from_file(mem_backend_path, mem_state, track_dirty_pages)
+                .map_err(RestoreFromSnapshotGuestMemoryError::File)?
+        }
+        MemBackendType::Uffd => {
+            return Err(RestoreFromSnapshotGuestMemoryError::UffdUnsupported.into());
         }
-        MemBackendType::Uffd => guest_memory_from_uffd(
-            mem_backend_path,
-            mem_state,
-            track_dirty_pages,
-            vm_resources.machine_config.huge_pages,
-        )
-        .map_err(RestoreFromSnapshotGuestMemoryError::Uffd)?,
     };
     builder::build_microvm_from_snapshot(
         instance_info,
         event_manager,
         microvm_state,
         guest_memory,
-        uffd,
         seccomp_filters,
         vm_resources,
         params.clock_realtime,
@@ -504,7 +464,7 @@ pub enum GuestMemoryFromFileError {
     File(#[from] std::io::Error),
     /// Failed to restore guest memory: {0}
     Restore(#[from] MemoryError),
-    /// Cannot restore hugetlbfs backed snapshot by mapping the memory file. Please use uffd.
+    /// Cannot restore hugetlbfs backed snapshot by mapping the memory file.
     HugetlbfsSnapshot,
 }

@@ -518,135 +478,8 @@ fn guest_memory_from_file(
     Ok(guest_mem)
 }

-/// Error type for [`guest_memory_from_uffd`]
-#[derive(Debug, thiserror::Error, displaydoc::Display)]
-pub enum GuestMemoryFromUffdError {
-    /// Failed to restore guest memory: {0}
-    Restore(#[from] MemoryError),
-    /// Failed to UFFD object: {0}
-    Create(userfaultfd::Error),
-    /// Failed to register memory address range with the userfaultfd object: {0}
-    Register(userfaultfd::Error),
-    /// Failed to connect to UDS Unix stream: {0}
-    Connect(#[from] std::io::Error),
-    /// Failed to sends file descriptor: {0}
-    Send(#[from] vmm_sys_util::errno::Error),
-}
-
-fn guest_memory_from_uffd(
-    mem_uds_path: &Path,
-    mem_state: &GuestMemoryState,
-    track_dirty_pages: bool,
-    huge_pages: HugePageConfig,
-) -> Result<(Vec<GuestRegionMmap>, Option<Uffd>), GuestMemoryFromUffdError> {
-    let (guest_memory, backend_mappings) =
-        create_guest_memory(mem_state, track_dirty_pages, huge_pages)?;
-
-    let mut uffd_builder = UffdBuilder::new();
-
-    // We only make use of this if balloon devices are present, but we can enable it unconditionally
-    // because the only place the kernel checks this is in a hook from madvise, e.g. it doesn't
-    // actively change the behavior of UFFD, only passively. Without balloon devices
-    // we never call madvise anyway, so no need to put this into a conditional.
-    uffd_builder.require_features(FeatureFlags::EVENT_REMOVE);
-
-    let uffd = uffd_builder
-        .close_on_exec(true)
-        .non_blocking(true)
-        .user_mode_only(false)
-        .create()
-        .map_err(GuestMemoryFromUffdError::Create)?;
-
-    for mem_region in guest_memory.iter() {
-        uffd.register(mem_region.as_ptr().cast(), mem_region.size() as _)
-            .map_err(GuestMemoryFromUffdError::Register)?;
-    }
-
-    send_uffd_handshake(mem_uds_path, &backend_mappings, &uffd)?;
-
-    Ok((guest_memory, Some(uffd)))
-}
-
-fn create_guest_memory(
-    mem_state: &GuestMemoryState,
-    track_dirty_pages: bool,
-    huge_pages: HugePageConfig,
-) -> Result<(Vec<GuestRegionMmap>, Vec<GuestRegionUffdMapping>), GuestMemoryFromUffdError> {
-    let guest_memory = memory::anonymous(mem_state.regions(), track_dirty_pages, huge_pages)?;
-    let mut backend_mappings = Vec::with_capacity(guest_memory.len());
-    let mut offset = 0;
-    for mem_region in guest_memory.iter() {
-        #[allow(deprecated)]
-        backend_mappings.push(GuestRegionUffdMapping {
-            base_host_virt_addr: mem_region.as_ptr() as u64,
-            size: mem_region.size(),
-            offset,
-            page_size: huge_pages.page_size(),
-            page_size_kib: huge_pages.page_size(),
-        });
-        offset += mem_region.size() as u64;
-    }
-
-    Ok((guest_memory, backend_mappings))
-}
-
-fn send_uffd_handshake(
-    mem_uds_path: &Path,
-    backend_mappings: &[GuestRegionUffdMapping],
-    uffd: &impl AsRawFd,
-) -> Result<(), GuestMemoryFromUffdError> {
-    // This is safe to unwrap() because we control the contents of the vector
-    // (i.e GuestRegionUffdMapping entries).
-    let backend_mappings = serde_json::to_string(backend_mappings).unwrap();
-
-    let socket = UnixStream::connect(mem_uds_path)?;
-    socket.send_with_fd(
-        backend_mappings.as_bytes(),
-        // In the happy case we can close the fd since the other process has it open and is
-        // using it to serve us pages.
-        //
-        // The problem is that if other process crashes/exits, firecracker guest memory
-        // will simply revert to anon-mem behavior which would lead to silent errors and
-        // undefined behavior.
-        //
-        // To tackle this scenario, the page fault handler can notify Firecracker of any
-        // crashes/exits. There is no need for Firecracker to explicitly send its process ID.
-        // The external process can obtain Firecracker's PID by calling `getsockopt` with
-        // `libc::SO_PEERCRED` option like so:
-        //
-        // let mut val = libc::ucred { pid: 0, gid: 0, uid: 0 };
-        // let mut ucred_size: u32 = mem::size_of::<libc::ucred>() as u32;
-        // libc::getsockopt(
-        //      socket.as_raw_fd(),
-        //      libc::SOL_SOCKET,
-        //      libc::SO_PEERCRED,
-        //      &mut val as *mut _ as *mut _,
-        //      &mut ucred_size as *mut libc::socklen_t,
-        // );
-        //
-        // Per this linux man page: https://man7.org/linux/man-pages/man7/unix.7.html,
-        // `SO_PEERCRED` returns the credentials (PID, UID and GID) of the peer process
-        // connected to this socket. The returned credentials are those that were in effect
-        // at the time of the `connect` call.
-        //
-        // Moreover, Firecracker holds a copy of the UFFD fd as well, so that even if the
-        // page fault handler process does not tear down Firecracker when necessary, the
-        // uffd will still be alive but with no one to serve faults, leading to guest freeze.
-        uffd.as_raw_fd(),
-    )?;
-
-    // We prevent Rust from closing the socket file descriptor to avoid a potential race condition
-    // between the mappings message and the connection shutdown. If the latter arrives at the UFFD
-    // handler first, the handler never sees the mappings.
-    forget(socket);
-
-    Ok(())
-}
-
 #[cfg(test)]
 mod tests {
-    use std::os::unix::net::UnixListener;
-
     use vmm_sys_util::tempfile::TempFile;

     use super::*;
@@ -666,7 +499,6 @@ mod tests {
     use crate::vmm_config::balloon::BalloonDeviceConfig;
     use crate::vmm_config::net::NetworkInterfaceConfig;
     use crate::vmm_config::vsock::tests::default_config;
-    use crate::vstate::memory::{GuestMemoryRegionState, GuestRegionType};

     fn default_vmm_with_devices() -> Vmm {
         let mut event_manager = EventManager::new().expect("Cannot create EventManager");
@@ -765,65 +597,4 @@ mod tests {
         )
     }

-    #[test]
-    fn test_create_guest_memory() {
-        let mem_state = GuestMemoryState {
-            regions: vec![GuestMemoryRegionState {
-                base_address: 0,
-                size: 0x20000,
-                region_type: GuestRegionType::Dram,
-                plugged: vec![true],
-            }],
-        };
-
-        let (_, uffd_regions) =
-            create_guest_memory(&mem_state, false, HugePageConfig::None).unwrap();
-
-        assert_eq!(uffd_regions.len(), 1);
-        assert_eq!(uffd_regions[0].size, 0x20000);
-        assert_eq!(uffd_regions[0].offset, 0);
-        assert_eq!(uffd_regions[0].page_size, HugePageConfig::None.page_size());
-    }
-
-    #[test]
-    fn test_send_uffd_handshake() {
-        #[allow(deprecated)]
-        let uffd_regions = vec![
-            GuestRegionUffdMapping {
-                base_host_virt_addr: 0,
-                size: 0x100000,
-                offset: 0,
-                page_size: HugePageConfig::None.page_size(),
-                page_size_kib: HugePageConfig::None.page_size(),
-            },
-            GuestRegionUffdMapping {
-                base_host_virt_addr: 0x100000,
-                size: 0x200000,
-                offset: 0,
-                page_size: HugePageConfig::Hugetlbfs2M.page_size(),
-                page_size_kib: HugePageConfig::Hugetlbfs2M.page_size(),
-            },
-        ];
-
-        let uds_path = TempFile::new().unwrap();
-        let uds_path = uds_path.as_path();
-        std::fs::remove_file(uds_path).unwrap();
-
-        let listener = UnixListener::bind(uds_path).expect("Cannot bind to socket path");
-
-        send_uffd_handshake(uds_path, &uffd_regions, &std::io::stdin()).unwrap();
-
-        let (stream, _) = listener.accept().expect("Cannot listen on UDS socket");
-
-        let mut message_buf = vec![0u8; 1024];
-        let (bytes_read, _) = stream
-            .recv_with_fd(&mut message_buf[..])
-            .expect("Cannot recv_with_fd");
-        message_buf.resize(bytes_read, 0);
-
-        let deserialized: Vec<GuestRegionUffdMapping> =
-            serde_json::from_slice(&message_buf).unwrap();
-
-        assert_eq!(uffd_regions, deserialized);
-    }
 }
diff --git a/src/vmm/src/vstate/vm.rs b/src/vmm/src/vstate/vm.rs
index 4ba4bdc..65f37d0 100644
--- a/src/vmm/src/vstate/vm.rs
+++ b/src/vmm/src/vstate/vm.rs
@@ -20,7 +20,6 @@ use kvm_bindings::{
 };
 use kvm_ioctls::VmFd;
 use serde::{Deserialize, Serialize};
-use userfaultfd::Uffd;
 use vmm_sys_util::errno;
 use vmm_sys_util::eventfd::EventFd;
 use vmm_sys_util::terminal::Terminal;
@@ -74,8 +73,6 @@ pub struct VmCommon {
     pub mmio_bus: Arc<Bus>,
     /// The global KVM state (fd + capabilities).
     pub kvm: Kvm,
-    /// Userfaultfd kept open for snapshot restore.
-    pub uffd: Option<Uffd>,
     /// Handles to vCPU threads.
     pub vcpus_handles: Mutex<Vec<VcpuHandle>>,
     /// Event fd written to by vCPUs on exit.
@@ -185,7 +182,6 @@ impl KvmVm {
             resource_allocator: Mutex::new(ResourceAllocator::new()),
             mmio_bus: Arc::new(Bus::new()),
             kvm,
-            uffd: None,
             vcpus_handles: Mutex::new(Vec::new()),
             vcpus_exit_evt,
         })
@@ -227,11 +223,6 @@ impl KvmVm {
         self.common.vcpus_handles.lock().expect("Poisoned lock")
     }

-    /// Sets the userfaultfd (used during snapshot restore).
-    pub fn set_uffd(&mut self, uffd: Option<Uffd>) {
-        self.common.uffd = uffd;
-    }
-
     /// Starts the microVM vCPUs.
     ///
     /// Sets the terminal to raw/non-blocking mode, then spawns a thread per vCPU
